### DISCUSSION TITLE
Unfinished string formatting in Windows Installer messages?

### DISCUSSION THREAD

**Discussion by symetritechadmin** [oleg-shilo/wixsharp#1504] [Q&A]
Hi!
When running an installation built with Wix4 / WixSharp 2.1.2 I'm seeing several messages in the standard Windows installer dialogs with what seems to be string placeholders not properly replaced. Some examples are:
 
Status: Searching for related applications Found application: [1]
Status: Removing applications Application: [1], Command line: [2]
Status: Copying new lines File: [1], Directory: [9], Size: [6]
Status: Generating script operations for action: [1]

What is this, is it by design or is it a formatting error somewhere?

Regards,
Lars

**Comment by oleg-shilo** [comment]
It's an interesting problem. 

If you have a look at the ProgressDialog code behind 
you will see that it always displays `messageRecord[2]`:

https://github.com/oleg-shilo/wixsharp/blob/a01cca3de29fe6d6aa20827d3a1e5ac4589521ef/Source/src/WixSharp.UI.WPF/Dialogs/ProgressDialog.xaml.cs#L167

In my experiments, it always was a clear text and I was never able to replicate the runtime conditions when item [2] was not clear text or was not present, which is something that I suspect you are experiencing. 

Without replicating it I could not confidently reverse engineer MSI progress message handling so I resorted to a simple algorithm just showing the record[2]. Even though I distribute a more comprehensive algorithm with WinForm version of ProgressDialog. Thus I disabled this algorithm until I find a way to test it:

https://github.com/oleg-shilo/wixsharp/blob/a01cca3de29fe6d6aa20827d3a1e5ac4589521ef/Source/src/WixSharp.UI/ManagedUI/Forms/ProgressDialog.cs#L147

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/d75fdefe-fcdd-4379-a641-db816d14e2b6)


Now, may I ask you a favour? Can you please copy and activate that strong algorithm (link above) in your ProgressDialog and check if it addresses the problem?

If it does, then I will update the oll VS templates with it.

**Reply by symetritechadmin** [reply]
Hi Oleg,
I finally got around to do some tests. I copied the ProgressDialog code to my code, renamed it, modified the dialog sequence accordingly and finally changed the "simple" value.
Some of the prompts change, but not all of them. I'm adding a couple of video clips with simple=true/false so you can see for yourself.
Regards, Lars

https://github.com/oleg-shilo/wixsharp/assets/145358325/f641f866-15b5-4857-acae-d538af465d01


https://github.com/oleg-shilo/wixsharp/assets/145358325/98b530f8-8643-47cc-b57f-9332375c9d3d

**Reply by symetritechadmin** [reply]
![image](https://github.com/oleg-shilo/wixsharp/assets/145358325/8c69197e-1419-40f2-b846-f5458097cbdc)

Debugger snapshot. It seems many of the messages are "nested", as if there would be more lines filling out the next level of [1], [2] etc. But there isn't. It all seems a bit strange.
Regards, Lars

**Reply by oleg-shilo** [reply]
@symetritechadmin , yep. 
That very last debugger screenshot brought me back ~5 years ago when I was reverse engineering this. While the parameters unfolding mechanism seems straightforward it was not supported by the runtime data.
Even in my code comments there I see the same artifacts that you are pointing out:

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/fd674ab9-17cd-44a5-901e-a301d8b0cebe)

And this is the reason why I opted for the "simple" algorithm. Because an alternative algorithm I no better.

Thus, sadly, I have no advice to give you. I have no idea how to obtain those additional formatting values (e.g. in your example `[1] and `[2]'`).

If I was in your shues, I would either leave it as is or put there a completely custom message based on the progress context analysis.

**Reply by symetritechadmin** [reply]
Yeah, when I looked at your code comments I realized that you must have come across this mess before 😅
I will leave it as it is for now. And as always, thanks for your rapid reply! 
Regards, Lars

**Reply by oleg-shilo** [reply]
> you must have come across this mess before

Mate, we are dealing with a 25-year-old technology. It is sometimes so painfully archaic that it's not even funny.
I created WixSharp to hide this ugliness from the people. But to do this I had to go through it all the way. I am scarred for life now 😃 

Just kidding. I am OK. But I am glad other people share my sentiment....

**Comment by madorjan** [comment]
This code snippet "fixed" the place-holder issue for me. Instead of templates with inserted values, the "non-template" versions of the progress texts are used.

WixSharp_wix4.bin (2.1.7)
ManagedUI (WinForms)

```
project.WixSourceGenerated += InjectUIRef;

....

private static void InjectUIRef(XDocument document)
{
    document.Root.Select("Package").AddElement("UIRef", "Id=WixUI_ErrorProgressText");
}
```
